src/ship_cmd.cpp

```diff
@@ -473,12 +473,27 @@ static Track ChooseShipTrack(Ship *v, TileIndex tile, DiagDirection enterdir, Tr
 		}
 		path_found = false;
 	} else {
+		/* Attempt to follow cached path. */
+		if (v->path_dest == v->dest_tile && !v->path.empty()) {
+			track = TrackdirToTrack(v->path.front());
+
+			if (HasBit(tracks, track)) {
+				v->path.pop_front();
+				v->HandlePathfindingResult(path_found);
```

Comment:
The cached path is chosen, even if it didn't find a valid path towards destination.
HandlePathfindingResult(true) may be wrong here.